### PR TITLE
on-demand definitions for compiler-built-in functions

### PR DESCRIPTION
**Pull request by kroening**
The definitions for the compiler-built-in functions are now generated on demand (whenever an undeclared function is encountered).

The time taken for 'make test' in regression/cbmc decreases from 268s to 63s.

### CODE REVIEW

File: src/ansi-c/builtin_factory.h
Comment (tautschnig) [inline src/ansi-c/builtin_factory.h:16]: So ... what does it do?

File: src/ansi-c/builtin_factory.cpp
Comment (tautschnig) [inline src/ansi-c/builtin_factory.cpp]: ?

File: src/ansi-c/Makefile
Comment (tautschnig) [inline src/ansi-c/Makefile]: Why is this order of file names better than the previous one?

File: src/ansi-c/builtin_factory.cpp
Comment (tautschnig) [inline src/ansi-c/builtin_factory.cpp]: Why is this here and not in `ansi_c_internal_additions.*`?

File: src/ansi-c/Makefile
Comment (tautschnig) [inline src/ansi-c/Makefile]: `builtin`, not `buitlin`

File: src/ansi-c/ansi_c_internal_additions.cpp
Comment (tautschnig) [inline src/ansi-c/ansi_c_internal_additions.cpp]: Why is this exempt from the move?

File: src/ansi-c/builtin_factory.cpp
Comment (tautschnig) [inline src/ansi-c/builtin_factory.cpp]: How about using a `std::istringstream` here, and thus getting rid of all the C code? Roughly:
````c++
std::istringstream hdr(header_file);
std::string line;
while(std::getline(hdr, line))
{
  line = strip_string(line);
  if(has_prefix(line, "//") || line.find(pattern) == std::string::npos)
    continue;
  
  out << line;
  return true;
}

return false;
````

**Review by tautschnig — CHANGES_REQUESTED** [review]

File: src/ansi-c/Makefile
Comment (kroening) [reply src/ansi-c/Makefile]: Now sorted

File: src/ansi-c/builtin_factory.h
Comment (kroening) [reply src/ansi-c/builtin_factory.h:16]: Comment added

File: src/ansi-c/builtin_factory.cpp
Comment (kroening) [reply src/ansi-c/builtin_factory.cpp]: Elaborated

File: src/ansi-c/Makefile
Comment (kroening) [reply src/ansi-c/Makefile]: fixed

File: src/ansi-c/ansi_c_internal_additions.cpp
Comment (kroening) [reply src/ansi-c/ansi_c_internal_additions.cpp]: done

File: src/ansi-c/builtin_factory.cpp
Comment (kroening) [reply src/ansi-c/builtin_factory.cpp]: moved

File: src/ansi-c/builtin_factory.cpp
Comment (kroening) [reply src/ansi-c/builtin_factory.cpp]: Will compare speed; the C++ version does memory allocation.

Comment (tautschnig) [reply src/ansi-c/builtin_factory.cpp]: Cool, thanks. If this is called frequently (which is where memory allocation cost would start to count) then there probably ought to be a cache (to be owned and cleared by the type checker) rather than parsing the very same strings again and again.

Comment (kroening) [reply src/ansi-c/builtin_factory.cpp]: I'll stick to the C version. The C++ above is 27x slower.

Comment (tautschnig) [reply src/ansi-c/builtin_factory.cpp]: As per our Chime conversation I think that by itself is not sufficient data to make a judgement upon. I would favour readability and maintainability over optimisation here, but I don't intend to make that a blocker.

**Review by tautschnig — APPROVED** [review]

File: src/ansi-c/CMakeLists.txt
Comment (tautschnig) [inline src/ansi-c/CMakeLists.txt]: Duplicate?

File: src/ansi-c/Makefile
Comment (tautschnig) [inline src/ansi-c/Makefile:65]: While at this, would it be possible to introduce a Makefile variable holding all the header files? As is, the same list of files appears three times.

File: src/ansi-c/ansi_c_internal_additions.cpp
Comment (tautschnig) [inline src/ansi-c/ansi_c_internal_additions.cpp:98]: Should this have a file name as the others do?

File: src/ansi-c/builtin_factory.h
Comment (tautschnig) [reply src/ansi-c/builtin_factory.h:16]: Not actually committed/pushed?

File: src/ansi-c/ansi_c_internal_additions.cpp
Comment (kroening) [reply src/ansi-c/ansi_c_internal_additions.cpp:98]: Seemed unnecessary luxury for two lines.

File: src/ansi-c/CMakeLists.txt
Comment (kroening) [reply src/ansi-c/CMakeLists.txt]: fixed

File: src/ansi-c/Makefile
Comment (kroening) [reply src/ansi-c/Makefile:65]: done

File: src/ansi-c/ansi_c_internal_additions.cpp
Comment (tautschnig) [reply src/ansi-c/ansi_c_internal_additions.cpp:98]: One day those 2 lines will be 10 and no-one will notice that they should have inserted the file name. It's near zero cost right now, but it might prove costly later.

**Review by chrisr-diffblue — APPROVED** [review]
Looks good to me, and will be great to see the speedup in running regression tests!

**Comment by chrisr-diffblue** [comment]
Oh, and I assume you'll be squashing together a few of the commits before merging too.

File: src/ansi-c/Makefile
Comment (tautschnig) [inline src/ansi-c/Makefile]: The above two are no longer necessary.

File: src/ansi-c/builtin_factory.cpp
Comment (tautschnig) [inline src/ansi-c/builtin_factory.cpp]: No longer necessary.

**Comment by kroening** [comment]
Will merge once CIed.

**Pull request merged**
